### PR TITLE
feat(mls): validate keyPackages count via MLS Client (AR-2284)

### PR DESCRIPTION
**Pull request by mchenani**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
Added a validKeyPackagesCount via MLSClient to bypass the 24 hours check limit.

### Issues

We were checking the package count every 24hours, but what if some how all of our packages drain before 24 hours? Also the API call has a cost! so with validKeyPackages count in mlsClient we can check the number without apiCalls.

With this changes, we check the valid key packages count via mlsClient, and if it's lower than the threshold, we bypass the 24 hours threshold and refill the keypackages.

### Causes (Optional)

Edge cases.

### Solutions

Implemented the solution.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/KeyPackageManager.kt [hunk 1]
```diff
@@ -55,7 +56,14 @@ internal class KeyPackageManagerImpl(
 
     private suspend fun refillKeyPackagesIfNeeded() {
         keyPackageRepository.value.lastKeyPackageCountCheck().flatMap { timestamp ->
-            if (Clock.System.now().minus(timestamp) > KEY_PACKAGE_COUNT_CHECK_DURATION) {
+            val forceRefill = keyPackageCountUseCase.value().let {
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/keypackage/KeyPackageManager.kt:59]: Shouldn't we pass `fromAPI` false here? 

```suggestion
            val forceRefill = keyPackageCountUseCase.value(fromAPI = false).let {
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/keypackage/KeyPackageLimitsProvider.kt [hunk 2]
```diff
@@ -15,9 +18,16 @@ class KeyPackageLimitsProviderImpl(
...
-    override val keyPackageUploadThreshold: Float
+    private val keyPackageUploadThreshold: Float
         get() = KEY_PACKAGE_THRESHOLD
 
+    override fun needsRefill(keyPackageCount: Int): Boolean {
+        return keyPackageCount < (keyPackageUploadLimit * keyPackageUploadThreshold)
+    }
+
+    override fun refillAmount(currentAmount: Int) =
+        keyPackageUploadLimit - currentAmount
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/keypackage/KeyPackageLimitsProvider.kt:29]: The use case changed so we should always refill the whole `keyPackageUploadLimit` now. See https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/556499553/Use+case+refill+key+packages+MLS#2.-Upload-new-key-packages for details.

**Review by typfel — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/keypackage/KeyPackageLimitsProvider.kt (hunk 2 above)
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/keypackage/KeyPackageLimitsProvider.kt]: Thanks, just kept the function as function to make it easy to modify/add logic later.

**Review by typfel — APPROVED** [review]
